src/main/java/com/google/devtools/build/remote/client/RemoteClient.java

```diff
@@ -118,14 +135,130 @@ private void listOutputDirectory(OutputDirectory dir, int limit) throws IOExcept
-  private void listTree(Path path, Tree tree, int limit) throws IOException, InterruptedException {
+  private void listTree(Path path, Tree tree, int limit) throws IOException {
     Map<Digest, Directory> childrenMap = new HashMap<>();
     for (Directory child : tree.getChildrenList()) {
       childrenMap.put(digestUtil.compute(child), child);
     }
     listDirectory(path, tree.getRoot(), childrenMap, limit);
   }
 
+  private static int getNumFiles(Tree tree) {
+    return tree.getChildrenList().stream().mapToInt(dir -> dir.getFilesCount()).sum();
+  }
+
+  // Outputs a bash executable line that corresponds to executing the given command.
+  private static void printCommand(Command command) {
+    for (EnvironmentVariable var : command.getEnvironmentVariablesList()) {
+      System.out.printf("%s=%s \\\n", var.getName(), ShellEscaper.escapeString(var.getValue()));
+    }
+    System.out.print("  ");
+
+    System.out.println(ShellEscaper.escapeJoinAll(command.getArgumentsList()));
+  }
+
+  private static void printList(List<String> list, int limit) {
+    if (list.isEmpty()) {
+      System.out.println("(none)");
+      return;
+    }
+    list.stream().limit(limit).forEach(name -> System.out.println(name));
+    if (list.size() > limit) {
+      System.out.println(" ... (too many to list, some omitted)");
+    }
+  }
+
+  // Output for print action command.
+  private void printAction(Action action, int limit) throws IOException {
+    Command command;
+    try {
+      command = Command.parseFrom(cache.downloadBlob(action.getCommandDigest()));
+    } catch (IOException e) {
+      throw new IOException("Could not obtain Command from digest.", e);
+    }
+    System.out.printf("Command [digest: %s]:\n", digestUtil.toString(action.getCommandDigest()));
+    printCommand(command);
+
+    Tree tree = cache.getTree(action.getInputRootDigest());
+    System.out.printf(
+        "\nInput files [total: %d, root Directory digest: %s]:\n",
+        getNumFiles(tree), digestUtil.toString(action.getCommandDigest()));
+    listTree(Paths.get(""), tree, limit);
+
+    System.out.println("\nOutput files:");
+    printList(action.getOutputFilesList(), limit);
+
+    System.out.println("\nOutput directories:");
+    printList(action.getOutputDirectoriesList(), limit);
+
+    System.out.println("\nPlatform:");
+    if (action.hasPlatform() && !action.getPlatform().getPropertiesList().isEmpty()) {
+      System.out.println(action.getPlatform().toString());
+    } else {
+      System.out.println("(none)");
+    }
+  }
+
+  // Display output file (either digest or raw bytes).
+  private void printOutputFile(OutputFile file, boolean showRawOutputs) {
+    String contentString;
+    if (file.hasDigest()) {
+      contentString = "Content digest: " + digestUtil.toString(file.getDigest());
+    } else if (showRawOutputs) {
+      contentString =
+          String.format(
+              "Raw contents: '%s', size (bytes): %d",
```

Comment:
We could consider having something like --show_contents_size_limit=20000 which will show everything less than the provided value of bytes. In the next CL, though.

Reply:
sounds good to me.